src/boot/efi/stub.c

```diff
@@ -21,6 +22,14 @@
 /* magic string to find in the binary image */
 _used_ _section_(".sdmagic") static const char magic[] = "#### LoaderInfo: systemd-stub " GIT_VERSION " ####";
 
+struct addon_entry {
+        char16_t        *source_path;
+        const EFI_DEVICE_PATH *device_path;
+};
+static int cmp_addon_entry(const struct addon_entry *s1, const struct addon_entry *s2) {
+        return strcmp16((const char16_t*)&s1->source_path, (const char16_t*)&s2->source_path);
```

Comment:
please compare by basename, not by path

Reply:
I'd see this problem as the "other face" of the problem mentioned regarding multiple initrds ordering anyway: https://github.com/systemd/systemd/pull/28070#issuecomment-1596761524

Maybe, all of this can be sidestepped by comparing for `.fname` section if it exist?

Reply:
well, i'd suppress all add-ons which carry a .fname where the actual filename doesn't match the .fname data. for all add-ons that pass this (i.e. either have a matching filename or don't have a PE section .fname) we should order by filename.

Reply:
Argh. Stop adding new sections for ever little thing. *every* section gets padded to 512 bytes on disk and to a full page when loaded into memory. This is stupidly wasteful fore a simple file name.

Just declare a new structured data section for this kind of stuff.

Reply:
@bluca for add-ons the ELF package metadata doens't really make sense does it, because typically locally generated, not packaged.

Reply:
Biggest unresolved point is this one.

Reply:
I introduced c0250f4d5a5aaf57d7e13303bee073545f17c600 to attempt to tackle this, can you let me know if this is the right direction or not?

Reply:
I wonder if it should be a JSON structure so that it can be extended by external consumers.

Reply:
(but JSON parser in UEFI… meh… .o O (ASN.1? :D))

Reply:
> @bluca for add-ons the ELF package metadata doens't really make sense does it, because typically locally generated, not packaged.

I imagine we'll start using it for more things, like for UKIs. You are always going on and on and on about how we don't validate filenames for anything after all, so it will likely apply to absolutely everything :>

Reply:
but maybe @medhefgo has a point, if it's not intended to be for "public" consumption, but just for internal use and sanity check, perhaps we could just use some data structure instead of adding a new header section?